secrets/gcp: add documentation for impersonated accounts feature (#18519)
hashicorp/vault: website/content/docs/secrets/gcp.mdx

For more information on creating and managing static accounts, see the
[GCP secrets engine API docs][api] docs.

> ## Impersonated Accounts

Reviewer: Maybe we can link to GCP docs on impersonation somewhere in this paragraph? Perhaps to https://cloud.google.com/iam/docs/impersonating-service-accounts.

Author: Sounds like a good idea to me. Added here: https://github.com/hashicorp/vault/pull/18519/files#diff-d9726e27b0de244125c80192a0fef3936511f40caf6aa37885d1976a9bc9a92fR223

> Impersonated accounts are a way to generate a GCP access token that is granted

Reviewer: Suggest to link to the common section that explains access tokens and be consistent with the wording (GCP -> OAuth2) used throughout the doc.

```suggestion
Impersonated accounts are a way to generate a OAuth2 [access token](/docs/secrets/gcp#access-tokens) that is granted
```

the permissions and accesses of another given service account. These access
tokens do not have the same 10-key limit as service account keys do, yet they
retain their short-lived nature. By default, their TTL in GCP is 1 hour, but
this may be configured to be up to 12 hours

> [as per Google Cloud documentation.](https://cloud.google.com/iam/docs/create-short-lived-credentials-delegated#sa-credentials-oauth)

Reviewer: Thanks for adding this bit on how they can be configured to have a longer TTL 👍

### Examples

To configure a Vault role that impersonates the administrator on the Google
Cloud project with the cloud platform and compute scopes:

```shell-session
$ vault write gcp/impersonated-account/my-token-impersonate \
    service_account_email="projectAdmin@my-project.iam.gserviceaccount.com" \
    token_scopes="https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/compute" \
    ttl="6h"
```

## Usage

After the secrets engine is configured and a user/machine has a Vault token with
the proper permission, it can generate credentials. Depending on how the Vault role
was configured, you can generate OAuth2 tokens or service account keys.

### Access Tokens

> To generate OAuth2 tokens, read from `gcp/.../token`. The role must have been

Author: Path here was left over from an older deprecated style so I updated it.

> To generate OAuth2 tokens, read from `gcp/.../token`. The role must have been
> created as type `access_token` or must be an impersonated account:

Reviewer: ```suggestion
To generate OAuth2 [access tokens](https://cloud.google.com/docs/authentication/token-types#access),
read from the [`gcp/.../token`](/api-docs/secret/gcp#generate-secret-iam-service-account-creds-oauth2-access-token)
API. The roleset or static account must have been created with a [`secret_type`](/api-docs/secret/gcp#secret_type)
of `access_token` or must be an impersonated account:
```

@MilenaHC @robmonte - I think we can make it clear that all three "role" types support the same access token generation in this section via the suggestion above. That addresses the topic of the discussion thread. Let me know your thoughts.

Author: I think that was a great place to explain it for all 3, nice!

For more information on creating and managing static accounts, see the
[GCP secrets engine API docs][api] docs.

## Impersonated Accounts

> Impersonated accounts are a way to generate a OAuth2 [access token](/docs/secrets/gcp#access-tokens) that is granted

Reviewer: ```suggestion
Impersonated accounts are a way to generate an OAuth2 [access token](/docs/secrets/gcp#access-tokens) that is granted
```

the permissions and accesses of another given service account. These access
tokens do not have the same 10-key limit as service account keys do, yet they
retain their short-lived nature. By default, their TTL in GCP is 1 hour, but
this may be configured to be up to 12 hours as explained in Google's 

> [short-lived credentials documentation.](https://cloud.google.com/iam/docs/create-short-lived-credentials-delegated#sa-credentials-oauth)

Reviewer: Nit: I think it's most common to leave punctuation out of the link.

token_ttl             3599
```

> ```shell-session

Reviewer: Similar to my suggestion on the API docs, we might want to provide examples of requesting an access token using all three types (e.g., roleset, static-account, impersonated-account)?

Author: Added with the api-docs commit

$ vault read gcp/impersonated-account/my-token-impersonate/token

Key                Value
---                -----
expires_at_seconds    1671667844
token                 ya29.c.b0AT7lpjBRmO7ghBEyMV18evd016hq...
token_ttl             59m59s
```

This endpoint generates a non-renewable, non-revocable static OAuth2 access token

> with a lifetime of one hour, where `token_ttl` is given in seconds and the

Reviewer: ```suggestion
with a max lifetime of one hour, where `token_ttl` is given in seconds and the
```

### Service Account Keys

> To generate service account keys, read from `gcp/.../key`. The roleset or static

Author: Path here was left over from an older deprecated style so I updated it. (Same as other location)

bigquery.datasets.update
```

When using service account impersonation, Vault must have the following additional
permissions:

```text
# impersonation
roles/iam.serviceAccountTokenCreator
roles/iam.workloadIdentityUser
```

> 

Author: Not sure if I am providing this info correctly and if this is correct location to do so.

Reviewer 2: The formatting and location seem fine to me. I can't speak for the correctness of the required permissions. Is that was you were referring to?

Reviewer 3: I believe the permissions needed are `iam.serviceAccountTokenCreator` and `iam.serviceAccountKeyAdmin` (see [the acceptance tests configuration](https://github.com/hashicorp/vault-plugin-secrets-gcp/commit/18d1fa3756da73fac75ba48b1fd52ab6b1b238a3#diff-5ccaa7ae143e28068ecdb6a41a4526ede72645e725116decfb76e0b1952aad35R41-R58))

We _may_ be able to remove `iam.serviceAccountTokenCreator` from just looking at our IAM configuration that we use in our environments but I'm not positive about that and would need to test that to be sure.

Author: Ah right, I should probably explain where that second permission came from.

[This page](https://cloud.google.com/iam/docs/impersonating-service-accounts#allow-impersonation) under the section named "Workload Identity User", says that `roles/workloadIdentityUser` is needed to impersonate service accounts from GKE workloads. I've never personally used GKE so I am not sure if it's even relevant or usable through the GCP secret engine, but that's why I added it here first and then left a comment to follow up.

I will try to find out whether it has no use here, or perhaps you have experience with it Matt?

Reviewer 3: I actually think this section you added here can be removed and instead the section on [static account permissions](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#LL472-L487) can be updated to read:

```text
When using static accounts or impersonated accounts, Vault must have the following permissions at the service account level:
```

A couple reasons for this:
1. All of the above are listing GCP permissions but what you've added are GCP roles which aggregate permissions and can be more permissive than desirable by users.
2. For impersonated accounts, this plugin only needs to be able to [get/lookup a service account](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/18d1fa3756da73fac75ba48b1fd52ab6b1b238a3/plugin/impersonated_account.go#L113). This is already covered [above in the required permissions](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L463).
3. The key permission to all of this is that the plugin's configured GCP Service Account has `iam.serviceAccounts.getAccessToken` permission **on the service account to impersonate**. This can be at the project, folder, or service account level but we should push users towards declaring it at the service account level which looks like is [already done above](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L479-L480).

The only other thing I would maybe do is remove [the following reference](https://github.com/hashicorp/vault/blob/5cf01bda057142b15987c44d53aa8dadf83b868c/website/content/docs/secrets/gcp.mdx?plain=1#L475-L477) in hopes to reduce the likelihood of someone unknowingly adding `roles/iam.serviceAccountKeyAdmin` to the wrong service account:

```
# Service account key admin
# This is provided in the `roles/iam.serviceAccountKeyAdmin` role
iam.serviceAccounts.get
```

Author: Thanks for pointing that out, you're right the roles don't belong being listed there. I agree that simply repurposing the permissions of static accounts to include impersonated accounts is better and it makes it more concise. I think the note about serviceAccountKeyAdmin doesn't quite belong there either, plus it's already mentioned in the section below so I'll remove it.
